src/cascadia/TerminalSettingsModel/WslDistroGenerator.cpp

```diff
@@ -136,3 +143,183 @@ std::vector<Profile> WslDistroGenerator::GenerateProfiles()
 
     return profiles;
 }
+
+// Function Description:
+// - Create a list of Profiles for each distro listed in names.
+// - Skips distros that are utility distros for docker (see GH#3556)
+// Arguments:
+// - names: a list of distro names to turn into profiles
+// Return Value:
+// - the list of profiles we've generated.
+static std::vector<Profile> namesToProfiles(const std::vector<std::wstring>& names)
+{
+    std::vector<Profile> profiles;
+    for (const auto& distName : names)
+    {
+        if (distName.substr(0, std::min(distName.size(), DockerDistributionPrefix.size())) == DockerDistributionPrefix)
+        {
+            // Docker for Windows creates some utility distributions to handle Docker commands.
+            // Pursuant to GH#3556, because they are _not_ user-facing we want to hide them.
+            continue;
+        }
+
+        auto WSLDistro{ CreateDefaultProfile(distName) };
+
+        WSLDistro.Commandline(L"wsl.exe -d " + distName);
+        WSLDistro.DefaultAppearance().ColorSchemeName(L"Campbell");
+        WSLDistro.StartingDirectory(DEFAULT_STARTING_DIRECTORY);
+        WSLDistro.Icon(L"ms-appx:///ProfileIcons/{9acb9455-ca41-5af7-950f-6bca1bc9722f}.png");
+        profiles.emplace_back(WSLDistro);
+    }
+    return profiles;
+}
+
+// Function Description:
+// - Open the reg key the root of the WSL data, in HKCU\Software\Microsoft\Windows\CurrentVersion\Lxss
+// Arguments:
+// - <none>
+// Return Value:
+// - the HKEY if it exists and we can read it, else nullptr
+static HKEY openWslRegKey()

```

Comment:
You can't have these return the `wil::unique_hkey` straight up?

Reply:
+1 for returning RAII wrappers.

Reply:
My preferred API surface when smart wrappers is involved -- and this extends to WRL::ComPtr and all similar variants -- is this:

```
smart_wrapper_thing GetAThing(); // produce shared wrapper
thing GetAThing(); // produce dumb thing, see note 1
void TakeAThingOptional(thing*); // consume only -- non-owning and non-shared -- optional
void TakeAThing(thing&); // consume only -- non-owning and non-shared -- required
```

_Of course, ref vs pointer is a decision you can make for the specific type. `HKEY` is by nature optional._

Putting the `smart_wrapper` into the consuming contract forces the caller to have that type (bad) and implies some ownership semantics (dependent on the type).

(1) In general, a function that returns a smart wrapper is subject to the same restrictions as one that returns the thing directly (like `HKEY openWslRegKey`): it must ensure that if an exception is encountered before the smart wrapper is populated, it does not leak the resource. Registry functions don't _throw_, so the need for returning a smart wrapper is reduced. That makes `HKEY open...` _okay_ in my book so long as we make sure to cover the error conditions.